File: minigit-4/ts/src/merge-algo.ts
// Three-way merge with conflict detection

import { MergeConflict } from './types';

export interface MergeResult {
  content: string;
  hasConflict: boolean;
}

export function merge3Way(base: string | null, ours: string, theirs: string): MergeResult {
  // If base is null, treat as new file in both branches
  if (base === null) {
    if (ours === theirs) {
      return { content: ours, hasConflict: false };
    }
    // Both added with different content - conflict
    return {
      content: formatConflict(ours, theirs, 'HEAD', 'incoming'),
      hasConflict: true,
    };
  }

  // If one side is unchanged
  if (base === ours) {
    return { content: theirs, hasConflict: false };
  }
  if (base === theirs) {
    return { content: ours, hasConflict: false };
  }

  // Both sides changed - need to merge
  const baseLines = base.split('\n');
  const oursLines = ours.split('\n');
  const theirsLines = theirs.split('\n');

  const result = mergeLines(baseLines, oursLines, theirsLines);
  return result;
}

function mergeLines(base: string[], ours: string[], theirs: string[]): MergeResult {
  // Simple LCS-based merge
  const oursChanges = computeChanges(base, ours);
  const theirsChanges = computeChanges(base, theirs);

  const result: string[] = [];
  let hasConflict = false;

  let baseIdx = 0;
  let oursIdx = 0;
  let theirsIdx = 0;

  while (baseIdx < base.length || oursIdx < ours.length || theirsIdx < theirs.length) {
    const oursChange = oursChanges.get(baseIdx);
    const theirsChange = theirsChanges.get(baseIdx);

    if (baseIdx >= base.length) {
      // Past end of base - handle remaining additions
      if (oursIdx < ours.length && theirsIdx < theirs.length) {
        if (ours.slice(oursIdx).join('\n') === theirs.slice(theirsIdx).join('\n')) {
          result.push(...ours.slice(oursIdx));
        } else {
          // Conflict at end
          hasConflict = true;
          result.push('<<<<<<< HEAD');
          result.push(...ours.slice(oursIdx));
          result.push('=======');
          result.push(...theirs.slice(theirsIdx));
          result.push('>>>>>>> incoming');
        }
      } else if (oursIdx < ours.length) {
        result.push(...ours.slice(oursIdx));
      } else if (theirsIdx < theirs.length) {
        result.push(...theirs.slice(theirsIdx));
      }
      break;
    }

    if (!oursChange && !theirsChange) {
      // No changes - keep base line
      result.push(base[baseIdx]);
      baseIdx++;
      oursIdx++;
      theirsIdx++;
    } else if (oursChange && !theirsChange) {
      // Only ours changed
      if (oursChange.type === 'delete') {
        baseIdx++;
        theirsIdx++;
      } else if (oursChange.type === 'modify') {
        result.push(oursChange.newLine!);
        baseIdx++;
        oursIdx++;
        theirsIdx++;
      } else if (oursChange.type === 'insert') {
        result.push(oursChange.newLine!);
        oursIdx++;
      }
    } else if (!oursChange && theirsChange) {
      // Only theirs changed
      if (theirsChange.type === 'delete') {
        baseIdx++;
        oursIdx++;
      } else if (theirsChange.type === 'modify') {
        result.push(theirsChange.newLine!);
        baseIdx++;
        oursIdx++;
        theirsIdx++;
      } else if (theirsChange.type === 'insert') {
        result.push(theirsChange.newLine!);
        theirsIdx++;
      }
    } else {
      // Both changed - check for conflict
      if (oursChange!.type === theirsChange!.type) {
        if (oursChange!.type === 'delete') {
          // Both deleted - ok
          baseIdx++;
        } else if (oursChange!.newLine === theirsChange!.newLine) {
          // Both made same change - ok
          result.push(oursChange!.newLine!);
          baseIdx++;
          oursIdx++;
          theirsIdx++;
        } else {
          // Different changes - conflict
          hasConflict = true;
          result.push('<<<<<<< HEAD');
          result.push(oursChange!.newLine || '');
          result.push('=======');
          result.push(theirsChange!.newLine || '');
          result.push('>>>>>>> incoming');
          baseIdx++;
          oursIdx++;
          theirsIdx++;
        }
      } else {
        // Different change types - conflict
        hasConflict = true;
        const oursContent = oursChange!.type === 'delete' ? '' : oursChange!.newLine || base[baseIdx];
        const theirsContent = theirsChange!.type === 'delete' ? '' : theirsChange!.newLine || base[baseIdx];

        result.push('<<<<<<< HEAD');
        if (oursContent) result.push(oursContent);
        result.push('=======');
        if (theirsContent) result.push(theirsContent);
        result.push('>>>>>>> incoming');

        baseIdx++;
        if (oursChange!.type !== 'delete') oursIdx++;
        if (theirsChange!.type !== 'delete') theirsIdx++;
      }
    }
  }

  return { content: result.join('\n'), hasConflict };
}

interface Change {
  type: 'insert' | 'delete' | 'modify';
  baseIdx: number;
  newLine?: string;
}

function computeChanges(base: string[], modified: string[]): Map<number, Change> {
  const changes = new Map<number, Change>();

  // Simple diff - not a full Myers diff for simplicity
  const lcs = longestCommonSubsequence(base, modified);

  let baseIdx = 0;
  let modIdx = 0;
  let lcsIdx = 0;

  while (baseIdx < base.length || modIdx < modified.length) {
    if (lcsIdx < lcs.length && baseIdx < base.length && modIdx < modified.length) {
      if (base[baseIdx] === lcs[lcsIdx] && modified[modIdx] === lcs[lcsIdx]) {
        // Same line - no change
        baseIdx++;
        modIdx++;
        lcsIdx++;
      } else if (base[baseIdx] !== lcs[lcsIdx]) {
        // Line deleted from base
        changes.set(baseIdx, { type: 'delete', baseIdx });
        baseIdx++;
      } else {
        // Line added in modified
        changes.set(baseIdx, { type: 'insert', baseIdx, newLine: modified[modIdx] });
        modIdx++;
      }
    } else if (baseIdx < base.length) {
      // Remaining lines in base are deleted
      changes.set(baseIdx, { type: 'delete', baseIdx });
      baseIdx++;
    } else if (modIdx < modified.length) {
      // Remaining lines in modified are added
      changes.set(baseIdx, { type: 'insert', baseIdx, newLine: modified[modIdx] });
      modIdx++;
    }
  }

  return changes;
}

function longestCommonSubsequence(a: string[], b: string[]): string[] {
  const m = a.length;
  const n = b.length;
  const dp: number[][] = Array(m + 1)
    .fill(null)
    .map(() => Array(n + 1).fill(0));

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (a[i - 1] === b[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }

  // Backtrack
  const result: string[] = [];
  let i = m;
  let j = n;

  while (i > 0 && j > 0) {
    if (a[i - 1] === b[j - 1]) {
      result.unshift(a[i - 1]);
      i--;
      j--;
    } else if (dp[i - 1][j] > dp[i][j - 1]) {
      i--;
    } else {
      j--;
    }
  }

  return result;
}

function formatConflict(ours: string, theirs: string, oursLabel: string, theirsLabel: string): string {
  const lines: string[] = [];
  lines.push(`<<<<<<< ${oursLabel}`);
  lines.push(ours);
  lines.push('=======');
  lines.push(theirs);
  lines.push(`>>>>>>> ${theirsLabel}`);
  return lines.join('\n');
}

export function findMergeBase(
  repoRoot: string,
  sha1: string,
  sha2: string,
  getParents: (sha: string) => string[]
): string | null {
  // Find common ancestor using BFS from both commits
  const ancestors1 = new Set<string>();
  const queue1: string[] = [sha1];

  // Build ancestor set for sha1
  while (queue1.length > 0) {
    const current = queue1.shift()!;
    if (ancestors1.has(current)) continue;
    ancestors1.add(current);

    const parents = getParents(current);
    queue1.push(...parents);
  }

  // BFS from sha2 to find first common ancestor
  const visited = new Set<string>();
  const queue2: string[] = [sha2];

  while (queue2.length > 0) {
    const current = queue2.shift()!;
    if (visited.has(current)) continue;
    visited.add(current);

    if (ancestors1.has(current)) {
      return current;
    }

    const parents = getParents(current);
    queue2.push(...parents);
  }

  return null;
}

File: minigit-4/py/merge_algo.py
"""Three-way merge algorithm with conflict detection for Mini Git."""

from typing import List, Tuple, Optional
from dataclasses import dataclass

from diff_algo import lcs_diff


@dataclass
class MergeResult:
    """Result of a three-way merge."""
    content: str
    has_conflicts: bool
    conflict_markers: List[Tuple[int, int]]  # (start, end) line numbers of conflicts


def three_way_merge(base: str, ours: str, theirs: str,
                    our_label: str = 'HEAD',
                    their_label: str = 'branch') -> MergeResult:
    """
    Perform a three-way merge.

    Args:
        base: Common ancestor content
        ours: Our version (current branch)
        theirs: Their version (branch being merged)
        our_label: Label for our side in conflict markers
        their_label: Label for their side in conflict markers

    Returns:
        MergeResult with merged content and conflict information
    """
    base_lines = base.splitlines(keepends=False)
    our_lines = ours.splitlines(keepends=False)
    their_lines = theirs.splitlines(keepends=False)

    # Get diffs from base to each version
    # Use lcs_diff but extract just op and line
    our_diff = [(op, line) for op, line, _, _ in lcs_diff(base_lines, our_lines)]
    their_diff = [(op, line) for op, line, _, _ in lcs_diff(base_lines, their_lines)]

    # Build change maps: which base lines changed in each version
    our_changes = {}  # base_line_idx -> (is_deleted, new_lines)
    their_changes = {}

    base_idx = 0
    new_lines = []
    for op, line in our_diff:
        if op == ' ':
            if new_lines:
                # Lines were added before this context
                if base_idx > 0:
                    if base_idx - 1 in our_changes:
                        our_changes[base_idx - 1][1].extend(new_lines)
                    else:
                        our_changes[base_idx - 1] = (False, new_lines)
                else:
                    our_changes[-1] = (False, new_lines)
                new_lines = []
            base_idx += 1
        elif op == '-':
            our_changes[base_idx] = (True, [])
            base_idx += 1
        else:  # '+'
            new_lines.append(line)

    if new_lines:
        our_changes[base_idx - 1 if base_idx > 0 else -1] = (False, new_lines)

    base_idx = 0
    new_lines = []
    for op, line in their_diff:
        if op == ' ':
            if new_lines:
                if base_idx > 0:
                    if base_idx - 1 in their_changes:
                        their_changes[base_idx - 1][1].extend(new_lines)
                    else:
                        their_changes[base_idx - 1] = (False, new_lines)
                else:
                    their_changes[-1] = (False, new_lines)
                new_lines = []
            base_idx += 1
        elif op == '-':
            their_changes[base_idx] = (True, [])
            base_idx += 1
        else:
            new_lines.append(line)

    if new_lines:
        their_changes[base_idx - 1 if base_idx > 0 else -1] = (False, new_lines)

    # Merge the changes
    result_lines = []
    conflicts = []
    has_conflicts = False

    # Handle insertions before first line
    if -1 in our_changes and -1 in their_changes:
        _, our_inserts = our_changes[-1]
        _, their_inserts = their_changes[-1]
        if our_inserts != their_inserts:
            has_conflicts = True
            conflict_start = len(result_lines)
            result_lines.append(f'<<<<<<< {our_label}')
            result_lines.extend(our_inserts)
            result_lines.append('=======')
            result_lines.extend(their_inserts)
            result_lines.append(f'>>>>>>> {their_label}')
            conflicts.append((conflict_start, len(result_lines)))
        else:
            result_lines.extend(our_inserts)
    elif -1 in our_changes:
        result_lines.extend(our_changes[-1][1])
    elif -1 in their_changes:
        result_lines.extend(their_changes[-1][1])

    for i, base_line in enumerate(base_lines):
        our_change = our_changes.get(i)
        their_change = their_changes.get(i)

        if our_change is None and their_change is None:
            # No changes
            result_lines.append(base_line)
        elif our_change is not None and their_change is None:
            # Only our change
            our_deleted, our_inserts = our_change
            if not our_deleted:
                result_lines.append(base_line)
            result_lines.extend(our_inserts)
        elif our_change is None and their_change is not None:
            # Only their change
            their_deleted, their_inserts = their_change
            if not their_deleted:
                result_lines.append(base_line)
            result_lines.extend(their_inserts)
        else:
            # Both changed - check for conflict
            our_deleted, our_inserts = our_change
            their_deleted, their_inserts = their_change

            if our_deleted == their_deleted and our_inserts == their_inserts:
                # Same change - no conflict
                if not our_deleted:
                    result_lines.append(base_line)
                result_lines.extend(our_inserts)
            else:
                # Conflict!
                has_conflicts = True
                conflict_start = len(result_lines)
                result_lines.append(f'<<<<<<< {our_label}')
                if not our_deleted:
                    result_lines.append(base_line)
                result_lines.extend(our_inserts)
                result_lines.append('=======')
                if not their_deleted:
                    result_lines.append(base_line)
                result_lines.extend(their_inserts)
                result_lines.append(f'>>>>>>> {their_label}')
                conflicts.append((conflict_start, len(result_lines)))

    return MergeResult(
        content='\n'.join(result_lines) + ('\n' if result_lines else ''),
        has_conflicts=has_conflicts,
        conflict_markers=conflicts
    )


def merge_files(base_content: Optional[str], our_content: Optional[str],
                their_content: Optional[str],
                our_label: str = 'HEAD',
                their_label: str = 'branch') -> Tuple[str, bool]:
    """
    Merge file contents handling creation/deletion cases.

    Returns:
        (merged_content, has_conflict)
    """
    # Handle cases where file doesn't exist in one or more versions
    if base_content is None:
        # File added in one or both branches
        if our_content is None:
            return their_content or '', False
        if their_content is None:
            return our_content, False
        if our_content == their_content:
            return our_content, False
        # Both added differently - conflict
        return (
            f'<<<<<<< {our_label}\n{our_content}=======\n{their_content}>>>>>>> {their_label}\n',
            True
        )

    if our_content is None:
        # Deleted in ours
        if their_content is None or their_content == base_content:
            return '', False  # Deleted in both or unchanged in theirs
        # Modified in theirs, deleted in ours - conflict
        return (
            f'<<<<<<< {our_label}\n=======\n{their_content}>>>>>>> {their_label}\n',
            True
        )

    if their_content is None:
        # Deleted in theirs
        if our_content == base_content:
            return '', False  # Unchanged in ours
        # Modified in ours, deleted in theirs - conflict
        return (
            f'<<<<<<< {our_label}\n{our_content}=======\n>>>>>>> {their_label}\n',
            True
        )

    # Both versions exist - do three-way merge
    if our_content == their_content:
        return our_content, False

    if our_content == base_content:
        return their_content, False

    if their_content == base_content:
        return our_content, False

    # Both modified differently - need three-way merge
    result = three_way_merge(base_content, our_content, their_content,
                             our_label, their_label)
    return result.content, result.has_conflicts


def find_merge_base(sha1: str, sha2: str, repo_root=None) -> Optional[str]:
    """
    Find the merge base (common ancestor) of two commits.
    Uses a simple BFS approach.
    """
    from objects import read_object, Commit

    # Get all ancestors of sha1
    ancestors1 = set()
    to_visit = [sha1]
    while to_visit:
        sha = to_visit.pop(0)
        if sha in ancestors1:
            continue
        ancestors1.add(sha)
        try:
            obj_type, data = read_object(sha, repo_root)
            if obj_type == 'commit':
                commit = Commit.deserialize(data)
                to_visit.extend(commit.parents)
        except:
            pass

    # Find first ancestor of sha2 that's also an ancestor of sha1
    to_visit = [sha2]
    visited = set()
    while to_visit:
        sha = to_visit.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        if sha in ancestors1:
            return sha

        try:
            obj_type, data = read_object(sha, repo_root)
            if obj_type == 'commit':
                commit = Commit.deserialize(data)
                to_visit.extend(commit.parents)
        except:
            pass

    return None
